src/storage/src/source/postgres.rs

```diff
@@ -128,28 +126,292 @@ macro_rules! try_recoverable {
+}
+
+/// An internal struct held by the spawned tokio task
+struct PostgresTaskInfo {
+    source_id: GlobalId,
+    connector: PostgresSourceConnector,
+    /// Our cursor into the WAL
+    lsn: PgLsn,
+    metrics: PgSourceMetrics,
+    source_tables: HashMap<u32, PostgresTable>,
+    sender: Sender<InternalMessage>,
+    activator: SyncActivator,
+}
+
+impl SourceReader for PostgresSourceReader {
+    type Key = ();
+    type Value = Row;
+    // Postgres can produce deletes that cause retractions
+    type Diff = Diff;
+
+    fn new(
+        _source_name: String,
         source_id: GlobalId,
-        connector: PostgresSourceConnector,
-        metrics: &SourceBaseMetrics,
-    ) -> Self {
-        Self {
-            source_id,
+        _worker_id: usize,
+        _worker_count: usize,
+        consumer_activator: SyncActivator,
+        connector: ExternalSourceConnector,
+        _restored_offsets: Vec<(PartitionId, Option<MzOffset>)>,
+        _encoding: SourceDataEncoding,
+        metrics: SourceBaseMetrics,
+        _connector_context: ConnectorContext,
+    ) -> Result<Self, anyhow::Error> {
+        let connector = match connector {
+            ExternalSourceConnector::Postgres(pg) => pg,
+            _ => {
+                panic!("Postgres is the only legitimate ExternalSourceConnector for PostgresSourceReader")
+            }
+        };
+
+        let (dataflow_tx, dataflow_rx) = tokio::sync::mpsc::channel(10_000);
+
+        let task_info = PostgresTaskInfo {
+            source_id: source_id.clone(),
+            connector: connector.clone(),
+            /// Our cursor into the WAL
+            lsn: 0.into(),
+            metrics: PgSourceMetrics::new(&metrics, source_id),
             source_tables: HashMap::from_iter(
                 connector
                     .details
                     .tables
                     .iter()
                     .map(|t| (t.relation_id, t.clone())),
             ),
-            connector,
-            lsn: 0.into(),
-            metrics: PgSourceMetrics::new(metrics, source_id),
+            sender: dataflow_tx,
+            activator: consumer_activator,
+        };
+
+        task::spawn(
+            || format!("postgres_source:{}", source_id),
+            postgres_replication_loop(task_info),
+        );
+        Ok(Self {
+            receiver_stream: dataflow_rx,
+        })
+    }
+
+    // TODO(guswynn): use `next` instead of using a channel
+    fn get_next_message(
+        &mut self,
+    ) -> Result<NextMessage<Self::Key, Self::Value, Self::Diff>, SourceReaderError> {
+        // TODO(guswynn): consider if `try_recv` is better or the same as `now_or_never`
+        let ret = match self.receiver_stream.recv().now_or_never() {
+            Some(Some(InternalMessage::Value {
+                value,
+                diff,
+                lsn,
+                end,
+            })) => {
+                let offset: u64 = lsn.into();
```

Comment:
I think I didnt do that just to avoid adding the pg dep in `mz-dataflow-types`? if you think that is okay, then I can do that! I will eventually switch offset to u64 and avoid any unwraps at all

Reply:
I don't see why not!  imo it's a reasonable dependency!  That crate depends on `kafka-util`, which, in turn, depends on `rdkafka`